src/signalcorr.jl

```diff
@@ -61,34 +60,31 @@ where each column in the result will correspond to a column in `x`.
 
 The output is not normalized. See [`autocor!`](@ref) for a method with normalization.
 """
-function autocov!(r::RealVector, x::RealVector, lags::IntegerVector; demean::Bool=true)
+function autocov!(
+    r::AbstractVector,
+    x::AbstractVector,
+    lags::AbstractVector{<:Integer},
+    z=zeros(typeof(zero(eltype(x)) / 1), length(x));
+    demean::Bool=true
+    )
     lx = length(x)
     m = length(lags)
     length(r) == m || throw(DimensionMismatch())
     check_lags(lx, lags)
-
     T = typeof(zero(eltype(x)) / 1)
-    z::Vector{T} = demean ? x .- mean(x) : x
+    demean  ? z = x :   z .= x .- mean(x)
```

Comment:
The order is wrong. Moreover, `z` should be modified in both cases - currently a new variable (possibly of a different type) is created in the first branch.


```suggestion
    demean ? z .= x .- mean(x) : copyto!(z, x)
```

Maybe it would be simpler to just change this to

```suggestion
    z = x .- (demean ? mean(x) : zero(eltype(x)) / 1)
```
and not pass `z` as an argument. This will be type stable. Maybe one could check if `x` is already of the correct element type but then it still can't be guaranteed that `x` would be of the same type as the other branch.

Reply:
If ```z``` is not passed as an argument, that will result in repeated allocations [here](https://github.com/JuliaStats/StatsBase.jl/pull/742/files#diff-415d7dec9339a5c81ca5a48be3206acffd88d83bf8dc8ba685fef3e94880fb2fR87)

Reply:
I see, then you should use the first suggestion :slightly_smiling_face: Currently the allocations are not avoided and the implementation is incorrect.